inc/Config/QueryContext/HttpQueryContext.php

```diff
@@ -180,4 +242,49 @@ public function process_response( string $raw_response_data, array $input_variab
 	final public function is_response_data_collection(): bool {
 		return $this->output_schema['is_collection'] ?? false;
 	}
+
+	// @todo: consider splitting the datasource injection out from query context so we don't have to tie a query
+	// to a datasource when instantiating. instead, we can just require applying queries to datasources in query
+	// runner execution. ie: $query_runner->execute( $query, $datasource );
+	//
+	/** @psalm-suppress ParamNameMismatch reason: we want the clarity provided by the rename here */
+	final public static function from_array( array $config, ?ValidatorInterface $validator = null ): self|\WP_Error {
+		if ( ! isset( $config['datasource'] ) || ! $config['datasource'] instanceof HttpDatasourceInterface ) {
+			return new \WP_Error( 'missing_datasource', __( 'Missing datasource.', 'remote-data-blocks' ) );
+		}
+
+		$validator = $validator ?? new Validator( self::CONFIG_SCHEMA );
+		$validated = $validator->validate( $config );
+
+		if ( is_wp_error( $validated ) ) {
+			return $validated;
+		}
+
+		return new self( $config['datasource'], $config['input_schema'], $config['output_schema'], $config );
+	}
+
+	public function to_array(): array {
+		return [
+			'datasource'    => get_class( $this->get_datasource() ),
+			'input_schema'  => $this->input_schema,
+			'output_schema' => $this->output_schema,
+		];
+	}
+
+	/**
+	 * Replace placeholders in the endpoint url with matching input variables.
+	 * 
+	 * The placeholders are expected to be in the Ruby on Rails/Express.js/Phoenix/etc
+	 * placeholder format of :placeholder, eg, `/path/:variable_name`.
+	 */
+	protected function replace_endpoint_placeholders( string $endpoint, array $input_variables ): string {
```

Comment:
do we really need this? (maybe, but probably not yet). if we do need it i would prefer we delegate to a library or wordpress function or this is just gonna be a huge headache

Reply:
At two working days from release? Yes.

Again, I am approaching this as bridging two paradigms that are not yet aligned not something etched in stone.

Reply:
As mentioned above, I don't think this is a requirement for release